### ISSUE TITLE
calling 'onBeforeRequest' callback from nodejs side

### ISSUE THREAD

**Issue by disovi** [kyungw00k/nightmare-webrequest-addon#1]
I want to make some processing of 'details' on nodejs side and then decide whether I want to cancel that request or not. It is only possible to access details through 'onBeforeRequest' event now. But there is no access to 'callback' inside that event. Is it possible to add such functionality? If yes, can you give me a hint how to achieve this? thanks.

**Comment by kuraga** [comment]
ping

**Comment by kuraga** [comment]
@disovi @kyungw00k don't you know what is accessible in `listener` (`onBeforeRequest` callback)? What's the scope?

**Comment by disovi** [comment]
@kuraga scope is limited exact to function inner context. Listener function is stringified before being sent to electron.

**Comment by kuraga** [comment]
@disovi [here](https://github.com/kyungw00k/nightmare-webrequest-addon/blob/master/src/webrequest-on-before-send-headers.js#L17-L18) ? So, `this == {}`?

**Comment by disovi** [comment]
@kuraga unfortunately yes.

**Comment by kuraga** [comment]
@kyungw00k @disovi can we change `{}` to `{ 'win': win }`, etc.? Или мы не можем передавать эти объекты на ту сторону, @disovi ? If no, which objects could we point there?

**Comment by kyungw00k** [comment]
@disovi 

> I want to make some processing of 'details' on nodejs side and then decide whether I want to cancel that request or not.

As far as I know, if you want to cancel the specific request, you can filter the request by filtering url on `detail` object which passing to the first argument of the callback `callback` function.

For example, If you want to cancel the specific request, 

```js
const filter = {
  urls: ['https://*.github.com/*', '*://electron.github.io']
}

nightmare.onBeforeRequest(filter, function (details, callback) { /* Electron-side scope */
        /* POINT #1 */
        callback({cancel: true})
})
```

I guess the point of this question is handing `details` object with the NodeJS-side scope in `POINT #1`, right?

@kuraga 

> can we change {} to { 'win': win }, etc.?

Even if we change code `{}` to `{ 'win': win }`, it still under the electron-side scope.

The callback we provide in NodeJS-side is want to access `win` object in the Electron-side,
then it can be possible.

**Comment by kuraga** [comment]
@kyungw00k we want to have _read_ access to _a copy_ of NodeJS-side object.

My task is different compared to @disovi 's: I want to set HTTP headers from file content.

**Comment by kyungw00k** [comment]
@kuraga 

> we want to have read access to a copy of NodeJS-side object.

If your-NodeJS-side-object can be serializable, I think it can be access from Electron-side callback with a some modification of this addon.

But it would be the different issue from this issue, right?

**Comment by disovi** [comment]
> I guess the point of this question is handing details object with the NodeJS-side scope in POINT #1, right?

@kyungw00k you are right. The idea was to use adblock extension together with nightmare. Unfortunately electron [doesn't support](https://github.com/electron/electron/issues/1498#issuecomment-96995155) chrome extensions. But there is an npm package [Brave Adblock](https://github.com/brave/ad-block) which implements this functionality. So basically I wanted to combine this engine with nightmare.

**Comment by kuraga** [comment]
> But it would be the different issue from this issue, right?

@kyungw00k , right :smile: @disovi , sorry! I'll create an issue.

**Comment by kyungw00k** [comment]
@disovi 

>  But there is an npm package Brave Adblock which implements this functionality. 

I've been looking the part which of the code had implemented `the functionality` from the `Brave Adblock` project, but I couldn't. 

Could you do that for me?

**Comment by disovi** [comment]
@kyungw00k in their github-page sample: 
`var b1 = client.matches('http://www.brianbondy.com/public/ad/some-ad', FilterOptions.script, 'slashdot.org')`
The first param - url, the third one is the host name.
